main.py

```diff
@@ -26,9 +26,9 @@ def load_data(chart_id):
                                                 "message": "No such chart"}}
         return response, 404
 
-@app.route("/clustering/<algorithm>/<similarity>/<encoding>/<outlier>/<chart_id>")
-@app.route("/clustering/<algorithm>/<similarity>/<encoding>/<outlier>/<chart_id>/<key>")
-def cluster(algorithm, similarity, encoding, outlier, chart_id, key=None):
+@app.route("/clustering/<algorithm>/<similarity>/<encoding>/<outlier>/<rep>/<chart_id>")
+@app.route("/clustering/<algorithm>/<similarity>/<encoding>/<outlier>/<rep>/<chart_id>/<key>")
```

Comment:
Notice how long this route has become? I should have mentioned this earlier, but optional parameters should be handled as query parameters.

https://stackoverflow.com/questions/24892035/how-can-i-get-the-named-parameters-from-a-url-using-flask